samza-core/src/main/scala/org/apache/samza/system/StreamMetadataCache.scala

```diff
@@ -94,17 +112,120 @@ class StreamMetadataCache (
       case None => None
     }
   }
 
-  private def addToCache(systemStream: SystemStream, metadata: SystemStreamMetadata, now: Long) {
+  private def addToFullMetadataCache(systemStream: SystemStream, metadata: SystemStreamMetadata, now: Long) {
     lock synchronized {
-      cache += systemStream -> CacheEntry(metadata, now)
+      fullMetadataCache += systemStream -> CacheEntry(metadata, now)
+    }
+  }
+
+  /**
+    * Get the newest offset for the systemStreamPartition. This returns null if there is no newest offset.
+    * Returns a cached value if it is fresh enough, even if the cached value has no newest offset data (e.g. stream is
+    * empty). Otherwise, call the admin to get the newest offset.
+    * As an optimization, this will also prefetch the newest offsets for other systemStreamPartitions that have been
+    * fetched in previous calls to getNewestOffset.
+    * @param maxRetries Max number of retries to use when fetching the newest offset for an individual
+    *                   systemStreamPartition. This is also the number of retries used when doing prefetching for other
+    *                   systemStreamPartitions.
+    */
+  def getNewestOffset(systemStreamPartition: SystemStreamPartition, maxRetries: Integer): String = {
+    // needs synchronization so that other threads don't come in a try to also do a fetch
+    newestOffsetCache synchronized {
+      val now = clock.currentTimeMillis
+      getFromCache(newestOffsetCache, systemStreamPartition, now) match {
+        // has a fresh enough entry with data, return it
+        case Some((key, Some(offset))) => offset
+        // has a fresh enough entry which has no data, still don't want to fetch again yet
+        case Some((key, None)) => null
+        // no entry or stale entry; do some fetching
+        case None => fetchNewestOffsetWithPrefetching(systemStreamPartition, maxRetries, now)
+      }
+    }
+  }
+
+  /**
+    * Fetch the newest offset for requestedSystemStreamPartition.
+    * Also prefetch the newest offsets for any other stale SSPs which had the same system as
+    * requestedSystemStreamPartition.
+    */
+  private def fetchNewestOffsetWithPrefetching(requestedSystemStreamPartition: SystemStreamPartition,
+                                               maxRetries: Integer,
+                                               now: Long): String = {
+    val systemToFetch = requestedSystemStreamPartition.getSystem
+    /*
+     * If a systemStreamPartition is in the newestOffsetCache, then we have fetched it before, so include it in
+     * the metadata fetch request now if it is stale. Also make sure it has the same system, since we will only use the
+     * admin for the system of the requestedSystemStreamPartition.
+     */
+    var systemStreamPartitionsToFetch = newestOffsetCache
+      .filter({case (ssp, cacheEntry) => systemToFetch.equals(ssp.getSystem) && isStale(now, cacheEntry.lastRefreshMs)})
```

Comment:
Minor: Don't need () around {}. I.e. `filter { case (s, e) => ... }` should work